netket/vqs/exact/exact_expect.py

```diff
@@ -0,0 +1,114 @@
+from functools import partial, lru_cache
+from typing import Callable, Any, Tuple
+
+import numpy as np
+
+import jax
+from jax import numpy as jnp
+from jax import tree_map
+from netket import jax as nkjax
+from netket.stats import statistics, Stats
+from netket.utils.types import PyTree
+from netket.utils.dispatch import dispatch, TrueT, FalseT
+from netket.utils import mpi
+
+from netket.operator import (
+    DiscreteOperator,
+    AbstractSuperOperator,
+    Squared,
+)
+
+from .exact_state import ExactState
+
+
+def _check_hilbert(A, B):
+    if A.hilbert != B.hilbert:
+        raise NotImplementedError(  # pragma: no cover
+            f"Non matching hilbert spaces {A.hilbert} and {B.hilbert}"
+        )
+
+
+@lru_cache(5)
+def sparsify(Ô):
+    """
+    Converts to sparse but also cache the sparsificated result to speed up.
+    """
+    return Ô.to_sparse()
+
+
+@dispatch
+def expect(vstate: ExactState, Ô: DiscreteOperator) -> Stats:  # noqa: F811
+    _check_hilbert(vstate, Ô)
+
+    O = sparsify(Ô)
+    Ψ = vstate.to_array()
+
```

Comment:
```suggestion

# TODO: This performs the full computation on all MPI ranks. 
# It would be great if we could split the computation among ranks.
```